File: common/code/common/Cosmos.fs
namespace common

open Azure
open System
open Microsoft.Azure.Cosmos
open System.Net
open FSharpPlus
open FSharp.Control

type CosmosError =
    | AlreadyExists
    | NotFound
    | ETagMismatch

type CosmosId =
    private
    | CosmosId of string

    static member fromString value =
        match String.IsNullOrWhiteSpace value with
        | true -> Error "Cosmos ID cannot be null or whitespace."
        | false -> CosmosId value |> Ok

    static member generate() =
        Guid.CreateVersion7().ToString() |> CosmosId

    static member toString(CosmosId id) = id

type CosmosRecord<'a> =
    { Id: CosmosId
      ETag: ETag
      PartitionKey: PartitionKey
      Record: 'a }

type CosmosQueryOptions =
    { Query: QueryDefinition
      ContinuationToken: ContinuationToken option
      PartitionKey: PartitionKey option }

[<RequireQualifiedAccess>]
module CosmosQueryOptions =
    let fromQueryString query =
        { CosmosQueryOptions.Query = QueryDefinition(query)
          ContinuationToken = None
          PartitionKey = None }

    let setQueryString query options =
        { options with
            Query = QueryDefinition(query) }

    let setQueryParameter parameterName parameterValue options =
        { options with
            Query = options.Query.WithParameter(parameterName, parameterValue) }

[<RequireQualifiedAccess>]
module Cosmos =
    let getId json =
        monad {
            let! idString = JsonObject.getStringProperty "id" json

            match CosmosId.fromString idString with
            | Ok cosmosId -> return cosmosId
            | Error error -> return! JsonResult.failWithMessage error
        }

    let getETag json =
        monad {
            let! etagString = JsonObject.getStringProperty "_etag" json

            match ETag.fromString etagString with
            | Ok etag -> return etag
            | Error error -> return! JsonResult.failWithMessage error
        }

    let private getDocumentsFromResponseJson json =
        json
        |> bind JsonNode.asJsonObject
        |> bind (JsonObject.getJsonArrayProperty "Documents")
        |> bind JsonArray.getJsonObjects

    let private getDocumentsFromResponse (response: ResponseMessage) =
        async {
            let! node = JsonNode.fromStream response.Content

            return getDocumentsFromResponseJson node
        }

    let private getCurrentPageResults (iterator: FeedIterator) =
        async {
            match iterator.HasMoreResults with
            | true ->
                let! cancellationToken = Async.CancellationToken
                use! response = iterator.ReadNextAsync(cancellationToken) |> Async.AwaitTask
                let _ = response.EnsureSuccessStatusCode()

                let! documents = getDocumentsFromResponse response |> Async.map JsonResult.throwIfFail

                let continuationToken =
                    ContinuationToken.fromString response.ContinuationToken |> Result.toOption

                return (documents, continuationToken)
            | false -> return (Seq.empty, None)
        }

    let private getFeedIteratorResults (iterator: FeedIterator) =
        let generator (iterator: FeedIterator) =
            match iterator.HasMoreResults with
            | true -> getCurrentPageResults iterator |> map fst |> map (fun x -> Some(x, iterator))
            | false -> async.Return None

        iterator |> AsyncSeq.unfoldAsync generator |> AsyncSeq.concatSeq

    let private getFeedIterator (container: Container) (options: CosmosQueryOptions) =
        let queryDefinition = options.Query

        let continuationToken =
            options.ContinuationToken
            |> Option.map ContinuationToken.toString
            |> Option.toObj

        let queryRequestOptions =
            let requestOptions = QueryRequestOptions()

            options.PartitionKey
            |> Option.iter (fun partitionKey -> requestOptions.PartitionKey <- partitionKey)

            requestOptions

        container.GetItemQueryStreamIterator(queryDefinition, continuationToken, queryRequestOptions)

    let getQueryResults container query =
        getFeedIterator container query |> getFeedIteratorResults

    let createRecord (container: Container) partitionKey record =
        async {
            use stream = BinaryData.FromObjectAsJson(record).ToStream()

            let options =
                let options = ItemRequestOptions()
                options.IfNoneMatchEtag <- "*"
                options

            let! cancellationToken = Async.CancellationToken

            use! response =
                container.CreateItemStreamAsync(stream, partitionKey, options, cancellationToken)
                |> Async.AwaitTask

            match response.StatusCode with
            | HttpStatusCode.PreconditionFailed -> return Error CosmosError.AlreadyExists
            | _ ->
                let _ = response.EnsureSuccessStatusCode()
                return Ok()
        }

    let patchRecord (container: Container) partitionKey id eTag patchOperations =
        async {
            let! cancellationToken = Async.CancellationToken

            let id = CosmosId.toString id
            let operations = List.ofSeq patchOperations

            let options =
                let options = PatchItemRequestOptions()
                options.IfMatchEtag <- ETag.toString eTag
                options

            use! response =
                container.PatchItemStreamAsync(id, partitionKey, operations, options, cancellationToken)
                |> Async.AwaitTask

            match response.StatusCode with
            | HttpStatusCode.PreconditionFailed -> return Error CosmosError.ETagMismatch
            | _ ->
                let _ = response.EnsureSuccessStatusCode()
                return Ok()
        }

    let deleteRecord (container: Container) partitionKey id eTag =
        async {
            let! cancellationToken = Async.CancellationToken

            let id = CosmosId.toString id

            let options =
                let options = ItemRequestOptions()
                options.IfMatchEtag <- ETag.toString eTag
                options

            use! response =
                container.DeleteItemStreamAsync(id, partitionKey, options, cancellationToken)
                |> Async.AwaitTask


            match response.StatusCode with
            | HttpStatusCode.PreconditionFailed -> return Error CosmosError.ETagMismatch
            | _ ->
                let _ = response.EnsureSuccessStatusCode()
                return Ok()
        }

File: common/code/common/Cosmos.cs
using LanguageExt;
using LanguageExt.Common;
using LanguageExt.UnsafeValueAccess;
using Microsoft.Azure.Cosmos;
using OpenTelemetry;
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Net;
using System.Runtime.CompilerServices;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;

namespace common;

public sealed record CosmosId
{
    private readonly string value;

    private CosmosId(string value) => this.value = value;

    public override string ToString() => value;

    public static Fin<CosmosId> From(string value) =>
        string.IsNullOrWhiteSpace(value)
        ? Error.New("Cosmos ID cannot be null or whitespace.")
        : new CosmosId(value);

    public static CosmosId Generate() => new(Guid.CreateVersion7().ToString());
}

public sealed record CosmosRecord<T>
{
    public required CosmosId Id { get; init; }

    public required PartitionKey PartitionKey { get; init; }

    public required T Record { get; init; }

    public required ETag ETag { get; init; }
}

public sealed record CosmosQueryOptions
{
    public required QueryDefinition Query { get; init; }
    public Option<ContinuationToken> ContinuationToken { get; init; } = Option<ContinuationToken>.None;
    public Option<PartitionKey> PartitionKey { get; init; } = Option<PartitionKey>.None;
}

public record CosmosError : Expected
{
    protected CosmosError(string Message, int Code)
        : base(Message, Code, default) { }

    public sealed record AlreadyExists : CosmosError
    {
        private AlreadyExists()
            : base("Resource already exists.", 409) { }

        public static AlreadyExists Instance { get; } = new();
    }

    public sealed record NotFound : CosmosError
    {
        private NotFound()
            : base("Resource not found.", 404) { }

        public static NotFound Instance { get; } = new();
    }

    public sealed record ETagMismatch : CosmosError
    {
        private ETagMismatch()
            : base("ETag mismatch.", 412) { }

        public static ETagMismatch Instance { get; } = new();
    }
}

public static class CosmosModule
{
    public static JsonResult<CosmosId> GetCosmosId(JsonObject jsonObject) =>
        from id in jsonObject.GetStringProperty("id")
        from cosmosId in JsonResult.Lift(CosmosId.From(id))
        select cosmosId;

    public static JsonResult<ETag> GetETag(JsonObject jsonObject) =>
        from eTagString in jsonObject.GetStringProperty("_etag")
        from eTag in JsonResult.Lift(ETag.From(eTagString))
        select eTag;

    public static async IAsyncEnumerable<JsonObject> GetQueryResults(Container container,
                                                                     CosmosQueryOptions cosmosQueryOptions,
                                                                     [EnumeratorCancellation]
                                                                     CancellationToken cancellationToken)
    {
        using var iterator = GetFeedIterator(container, cosmosQueryOptions);

        await foreach (var result in GetQueryResults(iterator, cancellationToken))
        {
            yield return result;
        }
    }

    private static FeedIterator GetFeedIterator(Container container, CosmosQueryOptions cosmosQueryOptions)
    {
        var queryDefinition = cosmosQueryOptions.Query;
        var continuationToken = cosmosQueryOptions.ContinuationToken.ValueUnsafe()?.ToString();

        var queryRequestOptions = new QueryRequestOptions();
        cosmosQueryOptions.PartitionKey.Iter(partitionKey => queryRequestOptions.PartitionKey = partitionKey);

        return container.GetItemQueryStreamIterator(queryDefinition, continuationToken, queryRequestOptions);
    }

    private static async IAsyncEnumerable<JsonObject> GetQueryResults(FeedIterator iterator,
                                                                      [EnumeratorCancellation]
                                                                      CancellationToken cancellationToken)
    {
        Option<ContinuationToken> continuationToken;

        do
        {
            (var documents, continuationToken) = await GetCurrentPageResults(iterator, cancellationToken);
            foreach (var document in documents)
            {
                yield return document;
            }
        }
        while (continuationToken.IsSome);
    }

    private static async ValueTask<(
        ImmutableArray<JsonObject> Documents,
        Option<ContinuationToken> ContinuationToken
    )> GetCurrentPageResults(FeedIterator iterator, CancellationToken cancellationToken)
    {
        using var response = await iterator.ReadNextAsync(cancellationToken);

        response.EnsureSuccessStatusCode();

        var documents = await GetDocuments(response, cancellationToken);

        var continuationToken = ContinuationToken.From(response.ContinuationToken)
                                                 .ToOption();

        return (documents, continuationToken);
    }

    private static async ValueTask<ImmutableArray<JsonObject>> GetDocuments(ResponseMessage response,
                                                                            CancellationToken cancellationToken)
    {
        var result = from jsonNode in await JsonNodeModule.From(response.Content, cancellationToken: cancellationToken)
                     from jsonObject in jsonNode.AsJsonObject()
                     from documentsJsonArray in jsonObject.GetJsonArrayProperty("Documents")
                     from documents in documentsJsonArray.GetJsonObjects()
                     select documents;

        return result.ThrowIfFail();
    }

    public static async ValueTask<Either<CosmosError, Unit>> CreateRecord(Container container,
                                                                          JsonObject jsonObject,
                                                                          PartitionKey partitionKey,
                                                                          CancellationToken cancellationToken)
    {
        using var stream = JsonNodeModule.ToStream(jsonObject);

        using var response =
            await container.CreateItemStreamAsync(stream,
                                                  partitionKey,
                                                  new ItemRequestOptions
                                                  {
                                                      
                                                      IfNoneMatchEtag = ETag.All.ToString()
                                                  },
                                                  cancellationToken);

        switch (response.StatusCode)
        {
            case HttpStatusCode.Conflict or HttpStatusCode.PreconditionFailed:
                return CosmosError.AlreadyExists.Instance;
            default:
                response.EnsureSuccessStatusCode();
                return Unit.Default;
        }
    }

    public static async ValueTask<Either<CosmosError, Unit>> PatchRecord(Container container,
                                                                         CosmosId id,
                                                                         PartitionKey partitionKey,
                                                                         IEnumerable<PatchOperation> patchOperations,
                                                                         ETag eTag,
                                                                         CancellationToken cancellationToken)
    {
        using var response =
            await container.PatchItemStreamAsync(id.ToString(),
                                                 partitionKey,
                                                 [.. patchOperations],
                                                 new PatchItemRequestOptions
                                                 {
                                                     IfMatchEtag = eTag.ToString()
                                                 },
                                                 cancellationToken);

        switch (response.StatusCode)
        {
            case HttpStatusCode.PreconditionFailed:
                return CosmosError.ETagMismatch.Instance;
            case HttpStatusCode.NotFound:
                return CosmosError.NotFound.Instance;
            default:
                response.EnsureSuccessStatusCode();
                return Unit.Default;
        }
    }

    public static async ValueTask<Either<CosmosError, Unit>> DeleteRecord(Container container,
                                                                          CosmosId id,
                                                                          PartitionKey partitionKey,
                                                                          ETag eTag,
                                                                          CancellationToken cancellationToken)
    {
        using var response =
            await container.DeleteItemStreamAsync(id.ToString(),
                                                  partitionKey,
                                                  new ItemRequestOptions
                                                  {
                                                      IfMatchEtag = eTag.ToString()
                                                  },
                                                  cancellationToken);

        switch (response.StatusCode)
        {
            case HttpStatusCode.PreconditionFailed:
                return CosmosError.ETagMismatch.Instance;
            case HttpStatusCode.NotFound:
                return Unit.Default;
            default:
                response.EnsureSuccessStatusCode();
                return Unit.Default;
        }
    }

    public static OpenTelemetryBuilder ConfigureTelemetry(OpenTelemetryBuilder builder)
    {
        AppContext.SetSwitch("Azure.Experimental.EnableActivitySource", true);

        return builder.WithTracing(tracing => tracing.AddSource("Azure.Cosmos.Operation"));
    }

    public static PartitionKey GetPartitionKey(Order order) => GetPartitionKey(order.Id);

    public static PartitionKey GetPartitionKey(OrderId orderId) => new(orderId.ToString());
}
